### ISSUE TITLE
Add tests for ftp.IFTPShell

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2855]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#2855|
|Type|enhancement|
|Created|2007-10-23 09:25:55Z|

Tests are good. This is extracted from [#7212](https://github.com/twisted/twisted/issues/7212).

<details><summary>Searchable metadata</summary>

```
trac-id__2855 2855
type__enhancement enhancement
reporter__therve therve
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__ 
time__1193131555000000 1193131555000000
changetime__1194515518000000 1194515518000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve removed owner|
|-|-|

This is ready to review in ftpshell-tests-2855.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to @therve|
|-|-|

A better test API for ```createFile``` would probably have it taking the file contents as an argument instead of looking it up as an attribute on self.

```createDirectory``` documentation should probably talk about whether path is relative or absolute, whether it can contain multiple segments, etc.  Maybe same for ```createFile```.

There could be some more tests in the direction of test_openForReading and test_openForWriting - tests which use the returned object and then make sure the correct bytes are read/written.  The test coverage provided by IReadWriteTestsMixin is good, but there's no association between IReadWriteTestsMixin.createFileReader/Writer and the objects returned by self.shell.openForReading/Writing in the IFTPShellTestsMixin tests.  One possibility is to do an isinstance test on what openForReading/Writing ultimately results in, to make sure it is the same class as IReadWriteTestsMixin ends up testing.  This might require making IReadWriteTestsMixin and IFTPShellTestsMixin know more about each other.

FTPReadWriteTestCase.tearDown doesn't need to exist, I think.  Leaving it behind helps with debugging, and it doesn't use an appreciable amount of resources.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

(In [21536]) Change createFile, some docs.

Refs [#2855](https://github.com/twisted/twisted/issues/2855)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

(In [21539]) Use the shell instead of intantiating the objects directly.

Refs [#2855](https://github.com/twisted/twisted/issues/2855)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve removed owner|
|-|-|

Thanks for the review. For the main point about the objects, I used the shell in ```FTPReadWriteTestCase```, so we're sure that we used the same objects.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @therve|
|-|-|

The EISDIR case of errnoToFailure is untested.  (so are a few others, but they're not new in this branch)

All of the exception cases for FTPAnonymousShell.stat are untested.  Same for FTPAnonymousShell.list, although that code was mostly there already.

I wonder if the unused exceptions should just be deleted.  Nothing ever used or raised them... but some code still might have been importing them or using them for some (pointless) reason.  I guess it's fine.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

(In [21557]) Add some more tests.

Refs [#2855](https://github.com/twisted/twisted/issues/2855)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve removed owner|
|-|-|

I've added a bunch of tests, back to review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @therve|
|-|-|

TestConsumer doesn't work with streaming producers, might be worth mentioning in the docstring.  Also it doesn't document its attributes, and doesn't use ```streamingProducer``` at all.

The rest looks good.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve commented|
|-|-|

(In [21591]) Some tweaks

Refs [#2855](https://github.com/twisted/twisted/issues/2855)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set status to `closed`|
|-|-|

(In [21592]) Merge ftpshell-tests-2855

Author: therve
Reviewers: exarkun, jml, cablehead
Fixes [#2855](https://github.com/twisted/twisted/issues/2855)

Add tests to the twisted.protocols.ftp.IFTPShell interface, to test the
implementation in ftp itself, and to allow to test against other implementations
in the future (eg, vfs).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve commented|
|-|-|

(In [21623]) Revert r21592: regression under win32-py24

Refs [#2855](https://github.com/twisted/twisted/issues/2855)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve set status to `reopened`|
|-|-|

The error:
```
[FAIL]: twisted.test.test_ftp.FTPShellTestCase.test_accessNotFound

Traceback (most recent call last):
  File "c:\twistedbuildbot24\W32-full2.4-scmikes-select\Twisted\twisted\trial\unittest.py", line 302, in _eb
    raise self.failureException(output)
twisted.trial.unittest.FailTest: 
Expected: (<class twisted.protocols.ftp.FileNotFoundError at 0x01AE4480>,)
Got:
[Failure instance: Traceback: exceptions.WindowsError: [Errno 3] The system cannot find the path specified: 'c:\\twistedbuildbot24\\W32-full2.4-scmikes-select\\Twisted\\_trial_temp\\twisted.test.test_ftp\\FTPShellTestCase\\test_accessNotFound\\_wfzl2\\temp\\foo/*.*'
c:\twistedbuildbot24\W32-full2.4-scmikes-select\Twisted\twisted\trial\unittest.py:608:_run
c:\twistedbuildbot24\W32-full2.4-scmikes-select\Twisted\twisted\internet\defer.py:107:maybeDeferred
c:\twistedbuildbot24\W32-full2.4-scmikes-select\Twisted\twisted\internet\utils.py:144:runWithWarningsSuppressed
c:\twistedbuildbot24\W32-full2.4-scmikes-select\Twisted\twisted\test\test_ftp.py:1903:test_accessNotFound
--- <exception caught here> ---
c:\twistedbuildbot24\W32-full2.4-scmikes-select\Twisted\twisted\protocols\ftp.py:1566:access
c:\twistedbuildbot24\W32-full2.4-scmikes-select\Twisted\twisted\python\filepath.py:477:listdir
]
```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve removed owner<br>@therve set status to `new`|
|-|-|

The fix is in r21625. Again, a problem with error reporting.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set owner to @therve|
|-|-|

Windows sucks.  and Python on Windows sucks.  Maybe someday we can get an API with uniform error reporting.  Until then, +1, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve commented|
|-|-|

(In [21626]) Something smarter

Refs [#2855](https://github.com/twisted/twisted/issues/2855)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_17"></a>|@therve commented|
|-|-|

(In [21644]) That didn't work.

Refs [#2855](https://github.com/twisted/twisted/issues/2855)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_18"></a>|@therve set status to `closed`|
|-|-|

(In [21645]) Merge ftpshell-tests-2855-2

Author: therve
Reviewers: exarkun, jml, cablehead
Fixes [#2855](https://github.com/twisted/twisted/issues/2855)

Add tests to the twisted.protocols.ftp.IFTPShell interface, to test the
implementation in ftp itself, and to allow to test against other
implementations in the future (eg, vfs).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
